app/assets/javascripts/components/generic_object/main-custom-button-form.js

```diff
@@ -72,35 +70,56 @@ function mainCustomButtonFormController(API, miqService, $q, $http) {
     ];
 
     miqService.sparkleOn();
-    optionsPromise = API.options('/api/custom_buttons')
-      .then(getCustomButtonOptions)
-      .catch(miqService.handleFailure);
+    var dataPromise;
+    var optionsPromise = API.options('/api/custom_buttons')
+      .then(function(response) {
+        _.forEach(response.data.custom_button_types, function(name, id) {
+          vm.button_types.push({id: id, name: name});
+        });
+        if (vm.customButtonRecordId) {
+          vm.newRecord = false;
+          miqService.sparkleOn();
+          dataPromise = API.get('/api/custom_buttons/' + vm.customButtonRecordId + '?attributes=resource_action,uri_attributes')
```

Comment:
this won't work - you're setting `dataPromise` *after* you use it in `$q.all`, so we never wait for this one.

But.. looks like no part of this request depends on anything from the options request, right?

So, maybe something like...

```
var optionsPromise = API.options('/api/custom_buttons')
  .then(function(response) {
    _.forEach(response.data.custom_button_types, function(name, id) {
      vm.button_types.push({id: id, name: name});
    });
  });

if (vm.customButtonRecordId) {
  var dataPromise = API.get('/api/custom_buttons/' + vm.customButtonRecordId + '?attributes=resource_action,uri_attributes');

  $q.all([dataPromise, optionsPromise])
    .then(function(data) {
      return getCustomButtonFormData(data[0]);
    })
    .catch(miqService.handleFailure);
} else {
  vm.newRecord = true;
  vm.modelCopy = angular.copy( vm.customButtonModel );
}
```

?

Reply:
Also, since we're always waiting for *some* responses,
we should probably `sparkleOn` every time (not only with id), and `sparkleOff` in the big `$q.all` block.